src/boot/efi/console.c

```diff
@@ -8,61 +8,109 @@
 
 #define EFI_SIMPLE_TEXT_INPUT_EX_GUID &(EFI_GUID) EFI_SIMPLE_TEXT_INPUT_EX_PROTOCOL_GUID
 
-EFI_STATUS console_key_read(UINT64 *key, BOOLEAN wait) {
+/*
+ * Reading input from the console sounds like an easy task to do, but thanks to broken
+ * firmware it is actually a nightmare.
+ *
+ * There is a ConIn and TextInputEx API for this. Ideally we want to use TextInputEx,
+ * because that gives us Ctrl/Alt/Shift key state information. Unfortunately, it is not
+ * always available and sometimes just non-functional.
+ *
+ * On the other hand we have ConIn, where some firmware likes to just freeze on us
+ * if we call ReadKeyStroke on it.
+ *
+ * Therefore, we use WaitForEvent on both ConIn and TextInputEx (if available) along
+ * with a timer event. The timer ensures there is no need to call into functions
+ * that might freeze on us, while still allowing us to show a timeout counter.
+ */
+EFI_STATUS console_key_read(UINT64 *key, UINT64 timeout_ms) {
         static EFI_SIMPLE_TEXT_INPUT_EX_PROTOCOL *TextInputEx;
         static BOOLEAN checked;
         UINTN index;
         EFI_INPUT_KEY k;
         EFI_STATUS err;
+        EFI_EVENT timer;
+        EFI_EVENT events[3] = { ST->ConIn->WaitForKey };
+        UINTN event_count = 1;
 
         if (!checked) {
                 err = LibLocateProtocol(EFI_SIMPLE_TEXT_INPUT_EX_GUID, (VOID **)&TextInputEx);
                 if (EFI_ERROR(err))
                         TextInputEx = NULL;
 
+                err = uefi_call_wrapper(BS->CheckEvent, 1, TextInputEx->WaitForKeyEx);
+                if (err == EFI_INVALID_PARAMETER)
+                        /* The firmware pretends it talks this protocol, but it really doesn't. */
+                        TextInputEx = NULL;
+                else
+                        events[event_count++] = TextInputEx->WaitForKeyEx;
+
                 checked = TRUE;
         }
 
-        /* wait until key is pressed */
-        if (wait)
-                uefi_call_wrapper(BS->WaitForEvent, 3, 1, &ST->ConIn->WaitForKey, &index);
+        if (timeout_ms > 0) {
+                err = uefi_call_wrapper(BS->CreateEvent, 5, EVT_TIMER, 0, NULL, NULL, &timer);
+                if (EFI_ERROR(err)) {
+                        PrintError(L"Error creating timer event: %r\n", err);
+                        return err;
+                }
+
+                /* SetTimer expects 100ns units for some reason. */
```

Comment:
hmm?

Reply:
I mean that instead of making it something sensible like taking nano seconds, they made it take 100ns. So passing a 1 is 100ns (which I find a silly unit to use); you have to multiply by 10 for microseconds.

Reply:
ah, got it. if we have to convert anyway, even more reaason to unify on usec here too